### ISSUE TITLE
Inquiry Regarding PTv3 SemanticKITTI Configuration

### ISSUE THREAD

**Issue by jingyuting1128** [Pointcept/Pointcept#556]
Hi,Thank you very much for your contributions to PTv3—your work has been incredibly helpful to us. Below are some confusions we encountered while configuring the SemanticKITTI setup for PTv3, and we would greatly appreciate your insights. We have configured the file three times, with only differences in the transform pipeline. Due to variations in the number of GPUs used, we also adjusted the batch size and learning rate accordingly. We have two main questions:
Why does the mIoU performance decrease when using more GPUs?
What is the correct order of transforms for SemanticKITTI?
Additionally, we kindly ask you to check if there are any overall errors in our configuration code. We look forward to your reply!

1. First Configuration (Single NVIDIA RTX 3090)
This setup referenced PTv3's nuScenes configuration. Here is the final experimental result:
[2025-12-09 23:52:29,698 INFO test.py line 340 3260] Val result: mIoU/mAcc/allAcc 0.6930/0.7496/0.9275

_base_ = ["../_base_/default_runtime.py"]

# Training Configuration
batch_size = 3  # Compromise: 50% faster than batch_size=2, safer than batch_size=4
mix_prob = 0.8
empty_cache = True  # Clear GPU memory fragments
enable_amp = True
amp_dtype = "float16"
empty_cache_per_epoch = True

# Model Configuration - PTv3
model = dict(
    type="DefaultSegmentorV2",
    num_classes=19,
    backbone_out_channels=64,
    backbone=dict(
        type="PT-v3m1",
        in_channels=4,
        order=["z", "z-trans", "hilbert", "hilbert-trans"],
        stride=(2, 2, 2, 2),
        enc_depths=(2, 2, 2, 6, 2),
        enc_channels=(32, 64, 128, 256, 512),
        enc_num_head=(2, 4, 8, 16, 32),
        enc_patch_size=(1024, 1024, 1024, 1024, 1024),
        dec_depths=(2, 2, 2, 2),
        dec_channels=(64, 64, 128, 256),
        dec_num_head=(4, 4, 8, 16),
        dec_patch_size=(1024, 1024, 1024, 1024),
        mlp_ratio=4,
        qkv_bias=True,
        qk_scale=None,
        attn_drop=0.0,
        proj_drop=0.0,
        drop_path=0.3,
        shuffle_orders=True,
        pre_norm=True,
        enable_rpe=False,
        enable_flash=True,
        upcast_attention=False,
        upcast_softmax=False,
        cls_mode=False,
        pdnorm_bn=False,
        pdnorm_ln=False,
        pdnorm_decouple=True,
        pdnorm_adaptive=False,
        pdnorm_affine=True,
        pdnorm_conditions=("nuScenes", "SemanticKITTI", "Waymo"),
    ),
    criteria=[
        dict(
            type="CrossEntropyLoss",
            weight=[
                3.1557, 8.7029, 7.8281, 6.1354, 6.3161, 7.9937, 8.9704,
                10.1922, 1.6155, 4.2187, 1.9385, 5.5455, 2.0198, 2.6261,
                1.3212, 5.1102, 2.5492, 5.8585, 7.3929
            ],
            loss_weight=1.0,
            ignore_index=-1,
        ),
        dict(
            type="LovaszLoss",
            mode="multiclass",
            loss_weight=1.0,
            ignore_index=-1,
        ),
    ],
)

# Transform Configuration
train_transform = [
    # Geometric Augmentations
    dict(type="RandomRotate", angle=[-1, 1], axis="z", center=[0, 0, 0], p=0.5),
    dict(type="RandomScale", scale=[0.9, 1.1]),
    dict(type="RandomFlip", p=0.5),
    dict(type="RandomJitter", sigma=0.005, clip=0.02),
    # SphereCrop as insurance to limit extremely large samples
    # point_max=120000 is more flexible than 100k (for batch_size=2)
    dict(type="SphereCrop", point_max=120000, mode="random"),
    # GridSample
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
    ),
    # Convert to Tensor
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment"),
        feat_keys=("coord", "strength"),
    ),
]

val_transform = [
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(type="SphereCrop", point_max=120000, mode="random"),
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
        return_inverse=True,
    ),
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment", "origin_segment", "inverse"),
        feat_keys=("coord", "strength"),
    ),
]

test_transform = [
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(
        type="GridSample",
        grid_size=0.025,
        hash_type="fnv",
        mode="train",
        return_inverse=True,
    ),
]

# Dataset Configuration
data = dict(
    num_classes=19,
    ignore_index=ignore_index,
    names=class_names,
    train=dict(
        type=dataset_type,
        split="train",
        data_root=data_root,
        transform=train_transform,
        test_mode=False,
        ignore_index=ignore_index,
    ),
    val=dict(
        type=dataset_type,
        split="val",
        data_root=data_root,
        transform=val_transform,
        test_mode=False,
        ignore_index=ignore_index,
    ),
    test=dict(
        type=dataset_type,
        split="val",
        data_root=data_root,
        transform=test_transform,
        test_mode=True,
        test_cfg=test_cfg,
        ignore_index=ignore_index,
    ),
)

# Other Configurations
pin_memory = True
persistent_workers = True
seed = 42
save_path = "exp/semantic_kitti/ptv3_baseline1"
resume = None
weight = None
evaluate = False
test_only = False

2. Second Configuration (Two NVIDIA RTX 4090s)
This setup directly adopted the model from PTv3's nuScenes configuration and the transform pipeline from PTv2's SemanticKITTI configuration (no modifications). We set batch_size=8 and lr=0.002. Here is the result:
[2026-01-10 17:50:33,429 INFO test.py line 340 42608] Val result: mIoU/mAcc/allAcc 0.6652/0.7319/0.9096

# Transform Configuration - Official PTv2 Order
train_transform = [
    dict(type="RandomRotate", angle=[-1, 1], axis="z", center=[0, 0, 0], p=0.5),
    dict(type="RandomScale", scale=[0.9, 1.1]),
    dict(type="RandomFlip", p=0.5),
    dict(type="RandomJitter", sigma=0.005, clip=0.02),
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
    ),
    dict(type="PointClip", point_cloud_range=(-35.2, -35.2, -4, 35.2, 35.2, 2)),
    dict(type="SphereCrop", sample_rate=0.8, mode="random"),
    dict(type="SphereCrop", point_max=120000, mode="random"),
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment"),
        feat_keys=("coord", "strength"),
    ),
]

val_transform = [
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
        return_inverse=True,
    ),
    dict(type="PointClip", point_cloud_range=(-35.2, -35.2, -4, 35.2, 35.2, 2)),
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment", "origin_segment", "inverse"),
        feat_keys=("coord", "strength"),
    ),
]

test_transform = [
    dict(type="PointClip", point_cloud_range=(-35.2, -35.2, -4, 35.2, 35.2, 2)),
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(
        type="GridSample",
        grid_size=0.025,
        hash_type="fnv",
        mode="train",
        return_inverse=True,
    ),
]

3. Third Configuration (Four NVIDIA RTX 4090s)
This setup strictly follows PTv3's paper:
Page 10 explicitly states: "PTv3 operates effectively without the need for point clipping within a specific range."
Table 16 (Outdoor Configuration) does not include SphereCrop—PTv3 does not require explicit point number limits.
Thus, we removed PointClip and SphereCrop for SemanticKITTI. We set batch_size=12 and lr=0.002. Here is the result:
[2026-01-12 23:24:28,290 INFO test.py line 340 50261] Val result: mIoU/mAcc/allAcc 0.6653/0.7215/0.9238

# Transform Configuration - Paper Table 16 Standard (No PointClip/SphereCrop)
# Training Set Transform
train_transform = [
    # Paper Table 16: Only 4 augmentations
    dict(type="RandomRotate", angle=[-1, 1], axis="z", center=[0, 0, 0], p=0.5),
    dict(type="RandomScale", scale=[0.9, 1.1]),
    dict(type="RandomFlip", p=0.5),
    dict(type="RandomJitter", sigma=0.005, clip=0.02),
    # GridSample (Paper Configuration: grid_size=0.05)
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
    ),
    # Conversion
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment"),
        feat_keys=("coord", "strength"),
    ),
]

# Validation Set Transform
val_transform = [
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(
        type="GridSample",
        grid_size=0.05,
        hash_type="fnv",
        mode="train",
        return_grid_coord=True,
        return_inverse=True,
    ),
    dict(type="ToTensor"),
    dict(
        type="Collect",
        keys=("coord", "grid_coord", "segment", "origin_segment", "inverse"),
        feat_keys=("coord", "strength"),
    ),
]

# Test Set Transform
test_transform = [
    dict(type="Copy", keys_dict={"segment": "origin_segment"}),
    dict(
        type="GridSample",
        grid_size=0.025,
        hash_type="fnv",
        mode="train",
        return_inverse=True,
    ),
]

We look forward to your reply and also anticipate the early release of the official SemanticKITTI configuration file for PTv3. Thank you very much!

**Comment by zbole** [comment]
<img width="2434" height="787" alt="Image" src="https://github.com/user-attachments/assets/ad06e3de-1130-45d4-b4b1-1045aa481f0b" />
My training results on KITTI are disappointing.
